### PR TITLE
Fix default-values for multiple-select in blocks

### PR DESCRIPTION
**Pull request by Prokyonn**
| Q | A
| --- | ---
| Bug fix? | yes
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | fixes https://github.com/sulu/sulu/issues/4987
| License | MIT
| Documentation PR | sulu/sulu-docs#prnum

#### What's in this PR?

Fixes the default-values in blocks. See https://github.com/sulu/sulu/issues/4987

An additional `value` attribute is introduced to save the state of the `value`. Previously the `value` was saved in the properties (`this.props`), but the properties are updated after the `default-value` is set for the next `single_select`, thus the `value` is not up to date during the rendering of the other components.

Now the new `value` attribute is manually updated before the next component is rendered.

#### Why?

The default value was only working for the last component.

#### Todo

 - [ ] Tests
 - [x] Rebase to 2.1
 - [ ] Check Imagemap

### CODE REVIEW

**Comment by nnatter** [comment]
Reminder: Check if we need to do something similar for the `ImageMap` when merging this 🙂

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/FieldBlocks/FieldBlocks.js [hunk 1]
```diff
@@ -2,7 +2,8 @@
 import equals from 'fast-deep-equal';
 import jsonpointer from 'json-pointer';
 import React, {Fragment} from 'react';
-import {toJS} from 'mobx';
+import {observable, action, toJS} from 'mobx';
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/containers/FieldBlocks/FieldBlocks.js:5]: Sort alphabetically.

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/FieldBlocks/FieldBlocks.js [hunk 2]
```diff
@@ -12,11 +13,24 @@ import FieldRenderer from './FieldRenderer';
...
+        super(props);
+
+        this.value = this.props.value;
+    }
+
     componentDidUpdate(prevProps: FieldTypeProps<Array<BlockEntry>>) {
         const {defaultType, onChange, types, value} = this.props;
         const {types: oldTypes} = prevProps;
 
+        if (!equals(prevProps.value, value) ){
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/containers/FieldBlocks/FieldBlocks.js:30]: Remove the space between the two closing brackets

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/FieldBlocks/FieldBlocks.js [hunk 3]
```diff
@@ -57,6 +76,8 @@ export default class FieldBlocks extends React.Component<FieldTypeProps<Array<Bl
         const newValues = toJS(oldValues);
         jsonpointer.set(newValues[index], '/' + name, value);
 
+        this.updateValue(newValues);
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/containers/FieldBlocks/FieldBlocks.js:79]: Just for my understanding: This is absolutely necessary, right? The `componentDidUpdate` call you've also touched is not executed between multiple `onChange` calls, right?

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/FieldBlocks/tests/FieldBlocks.test.js [hunk 4]
```diff
@@ -898,6 +901,130 @@ test('Should call onFinish when the order of the blocks has changed', () => {
...
 
+test('Should set correct default values for multiple single_select in blocks', () => {
+    const formInspector = new FormInspector(new ResourceFormStore(new ResourceStore('test'), 'test'));
+
+    const types = {
+        default: {
+            title: 'Default',
+            form: {
+                position_center: {
+                    label: 'Position Left',
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/containers/FieldBlocks/tests/FieldBlocks.test.js:912]: Rename to "Position Center"? Confused me a bit at first :see_no_evil:

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/FieldBlocks/tests/FieldBlocks.test.js [hunk 5]
```diff
@@ -898,6 +901,130 @@ test('Should call onFinish when the order of the blocks has changed', () => {
...
+            defaultType="default"
+            formInspector={formInspector}
+            minOccurs={1}
+            onChange={changeSpy}
+            types={types}
+            value={observable([{type: 'default'}])}
+        />
+    );
+
+    fieldBlocks.find('Block').at(0).simulate('click');
```
Comment (danrot) [inline src/Sulu/Bundle/AdminBundle/Resources/js/containers/FieldBlocks/tests/FieldBlocks.test.js:1015]: Out of curiosity: Is that click really necessary? Shouldn't the call below happen anyway? Because if it doesn't, there might be some edge case, which might cause some weird behavior.

**Review by danrot — CHANGES_REQUESTED** [review]
I think that's a better solution than [the two I came up with](https://github.com/sulu/sulu/issues/4987#issuecomment-576152129) :+1:

So I am generally fine with that, still added a few minor issues resp. questions for better understanding.

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/FieldBlocks/FieldBlocks.js (hunk 3 above)
Comment (Prokyonn) [reply src/Sulu/Bundle/AdminBundle/Resources/js/containers/FieldBlocks/FieldBlocks.js:79]: Yep, we debugged through it and the `componentDidUpdate` is called only after the second `onChange` is called. Thus the `value` is not yet updated during the second `onChange` call.

File: src/Sulu/Bundle/AdminBundle/Resources/js/containers/FieldBlocks/tests/FieldBlocks.test.js (hunk 5 above)
Comment (Prokyonn) [reply src/Sulu/Bundle/AdminBundle/Resources/js/containers/FieldBlocks/tests/FieldBlocks.test.js:1015]: No the call does not happen. The blocks are collapsed and the rendering is only called after they are expanded.

Comment (danrot) [reply src/Sulu/Bundle/AdminBundle/Resources/js/containers/FieldBlocks/tests/FieldBlocks.test.js:1015]: Ok, when a new block is added, it is already opened anyway... So that should not be that bad. I guess the only thing that might not work as expected is if a new field is added to a block type, which is already used somewhere. The value would probably not be updated, until the block is expanded in the UI. But I think that this is an edge case, which does not hurt.

**Review by danrot — APPROVED** [review]
